### PR TITLE
Add telemetry for commands

### PR DESCRIPTION
**Pull request by aeisenberg**
This commit adds telemetry capturing for command execution. The data captured is only the command id.

We also capture errors thrown by any command execution.

This adds telemetry to an Azure application insights instance. Talk to me if you want access to it.

We still need a way to notify users that telemetry is being collected and a way to disable it.

Also:

* Extract commandRunner to its own file
* Add an option to log all telemetry commands to the console
* Add a popup (text to be determined) that requires users to accept
  telemetry before using.
* Convert telemetry to opt-in instead of opt-out
* Add a "canary" setting that users can use to opt-in to unreleased
  features.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/docs-content-dsp` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

### CODE REVIEW

**Comment by aeisenberg** [comment]
Ping @github/docs-content-dsp!

We are going to need to create some documentation about what we are and what we are _not_ collecting for telemetry. @sj is working on some of the wording. See our internal issue for more information on this.

**Comment by rachmari** [comment]
@aeisenberg docs first responder here. 👋 I've opened a tracking issue for add the documentation page that you mentioned above: https://github.com/github/docs-content/issues/3125

**Comment by aeisenberg** [comment]
@sj could you take a look at the `TELEMETRY.md` file?

And at some point, I can walk you through what the UX is going to be like.

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:150]: It is a little strange to me to only call this during testing.
I see that `listener` and `reporter` are tracked in `ctx.subscriptions` and disposed when the extension is deactivated, but perhaps we can make that more explicit.

What do you think about organising this file into a `Telemetry` class that extends `DisposableObject`, that contains `listener` and `reporter` fields that are created explicitly at construction, can be reset if necessary, and are tracked for disposal? Bit like what we do for the server clients. Then we only push the `Telemetry` object to `ctx.subscriptions`.

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:105]: Should we also listen for the top-level telemetry setting here?

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:134]:
```suggestion
          'Do we have your permission to collect anonymous usage statistics to help us improve CodeQL for VSCode? See [TELEMETRY.md](https://github.com/github/vscode-codeql/blob/main/.github/TELEMETRY.md)',
```

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:137]: Is this timeout enough for a reader to take in the message and decide? What if they decide to follow the link and read?

Shame that the extension won't start if this isn't resolved. What do others to? Is it possible to have it continue with initialisation (without telemetry)?

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:29]: Should we be checking the global telemetry setting too, as stated in the writeup?
For consistency, I suggest adding an `isTelemetryEnabled` helper to `config.ts`, which checks the relevant settings, and always using that. (We could even enforce it by query.)

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:37]: Move comment down to the relevant code.

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:104]: We have `ConfigListener` classes/interfaces to listen for changes to the CLI settings. Would that be useful here? Could we avoid the recursion between `registerListener` and `initializeTelemetry` by listening for config changes and resetting the telemetry reporting state when any of the relevant settings change?

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:129]: Use the `Setting` object here.

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:117]: Either pull the string key out into a constant, or create a helper method that takes `ctx` and a boolean.

File: extensions/ql-vscode/src/vscode-tests/no-workspace/telemetry.test.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/telemetry.test.ts:75]:
```suggestion
  it('should not initialize telemetry when global option disabled', async () => {
```

File: extensions/ql-vscode/src/vscode-tests/no-workspace/telemetry.test.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/telemetry.test.ts:152]:
```suggestion
  it('should set userOptIn to false when global setting changes', async () => {
```

**Review by adityasharad — COMMENTED** [review]
Nice work. The approach looks sensible and most of my comments are questions or digging into details for clarity.

One major suggested change involves tracking the telemetry reporting state (and the associated listeners and config changes) as an object of a dedicated class, which I hope will simplify the logic but may require some rewiring.

File: .github/TELEMETRY.md
Comment (aeisenberg) [reply .github/TELEMETRY.md:21]: Hmmm...if I add the word _setting_ here, then I should also add it in the previous paragraph and I think that's redundant.

File: extensions/ql-vscode/src/helpers.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/helpers.ts:78]: This creates an optional options argument. At one point, I had several other options, but I removed them. Maybe now, I can convert this back to a regular optional boolean argument.

File: extensions/ql-vscode/src/helpers.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/helpers.ts:84]: Now there are 3 return types. We need to handle `undefined`, which means that the user has not yet made a choice.

Comment (aeisenberg) [reply extensions/ql-vscode/src/helpers.ts:84]: I will update the jsdoc for this function.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:29]: No. The global telemetry value is checked implicitly in the `vscode-extension-telemetry` package.

Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:29]: I started working on a PR so that this package can do all the checking of both settings, but the maintainer is not responsive. I created a very simple PR as a trial balloon, and there was no response.

Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:29]: > Should we be checking the global telemetry setting too, as stated in the writeup?
> For consistency, I suggest adding an `isTelemetryEnabled` helper to `config.ts`, which checks the relevant settings, and always using that. (We could even enforce it by query.)

I don't think this is necessary since the check is a one-liner anyway.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:137]: It's possible that the 10s timeout is not long enough. But if they miss it the first time, they will be notified again the next time they restart.

The salesforce extension uses a modal dialog. Doing the same would make this whole section much simpler, though it would be a worse experience for the user. I'm a little torn.

Perhaps we could try out a modal dialog with internal users first and see how they react. It would be easy to switch back to non-modal.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:150]: I'll try this out.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:117]: sure

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:129]: sure

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:105]: We don't need to. This is handled by the package.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts:104]: Let me see if that makes sense. I'd originally thought of using ConfigListeners, but there was a bit of impedance mismatch (but I've forgotten why, now).

I'll see if I can use this to also manage disposables more cleanly.

**Comment by aeisenberg** [comment]
Note that the failing job is because the release workflow cannot be run on forks.

**Comment by aeisenberg** [comment]
@sj We are hoping to release this shortly. Could you take a look at the `TELEMETRY.md` file as well as the changes to `README.md` and make sure they are sufficient from your perspective?

File: TELEMETRY.md
Comment (sj) [inline TELEMETRY.md:17]: ~Would it be worth clarifying here that we **do not** collect any information about the nature of the CodeQL queries that are being run?~ You already do this further down. Excellent!

File: TELEMETRY.md
Comment (sj) [inline TELEMETRY.md:3]: "for a maximum of 30 days" is not consistent with what is in the text below

> Aggregated data of command identifiers and run times will be retained for a maximum of 180 days.

File: TELEMETRY.md
Comment (sj) [inline TELEMETRY.md:20]: That's not consistent with the text above:

> This data will not be shared with any parties outside of GitHub and will be retained for a maximum of 30 days.

File: TELEMETRY.md
Comment (sj) [inline TELEMETRY.md:26]: Excellent!

File: TELEMETRY.md
Comment (sj) [inline TELEMETRY.md:37]: Here it is worth repeating again that telemetry reporting is disabled until and unless the user explicitly agrees with it.

File: extensions/ql-vscode/README.md
Comment (sj) [inline extensions/ql-vscode/README.md:116]: > and will be retained for a maximum of 30 days

There's a bit of an inconsistency here (see my other comment).

Comment (sj) [inline extensions/ql-vscode/README.md:116]: As per my other comments, it'd be great if we could work in the word "anonymized" a few times :slightly_smiling_face:

**Review by sj — CHANGES_REQUESTED** [review]
Overall looks :+1: with a few suggestions

File: TELEMETRY.md
Comment (aeisenberg) [reply TELEMETRY.md:3]: On further reflection, this is a bit of a problem. IP Address and the machine ID is not "anonymous". And we treat anonymous data differently. I'll have to clarify that.

File: TELEMETRY.md
Comment (aeisenberg) [reply TELEMETRY.md:13]: Good point.

**Comment by aeisenberg** [comment]
Thanks for your comments, @sj. I think I addressed them all. The wording is more complex now, but also more precise sine I need to distinguish between the anonymous and non-anonymized data.

**Comment by sj** [comment]
`TELEMETRY.md` and `README.md` look great, but I'm wondering whether one of your commits added a whole bunch of other unrelated changes to this PR. I'll leave that up to @adityasharad.

**Review by sj — APPROVED** [review]

**Comment by aeisenberg** [comment]
Thanks @sj! Yes, this PR includes all of the telemetry code for the extension.

**Comment by aeisenberg** [comment]
Also, the failing `Release` workflow is expected since it requires secrets specified in the main repository (not my fork).

File: TELEMETRY.md
Comment (adityasharad) [inline TELEMETRY.md:15]:
```suggestion
- Randomly generated GUID that uniquely identifies a CodeQL extension installation. (Discarded before aggregation.)
```

File: TELEMETRY.md
Comment (adityasharad) [inline TELEMETRY.md:31]:
```suggestion
- No CodeQL database names or contents
```

File: TELEMETRY.md
Comment (adityasharad) [inline TELEMETRY.md:32]:
```suggestion
- No contents of CodeQL queries
```

File: extensions/ql-vscode/README.md
Comment (adityasharad) [inline extensions/ql-vscode/README.md:116]:
```suggestion
If you specifically opt-in to permit GitHub to do so, GitHub will collect usage data and metrics for the purposes of helping the core developers to improve the CodeQL extension for VS Code. This data will not be shared with any parties outside of GitHub. IP addresses and installation IDs will be retained for a maximum of 30 days. Anonymous data will be retained for a maximum of 180 days. Please see [telemetry](TELEMETRY.md) for more information.
```

File: .github/workflows/release.yml
Comment (adityasharad) [inline .github/workflows/release.yml:100]: Why are all these changes here?

File: extensions/ql-vscode/src/telemetry.ts
Comment (adityasharad) [inline extensions/ql-vscode/src/telemetry.ts:148]: @sj this matches the first line of `TELEMETRY.md`: are you happy with this as the user-facing prompt?
```suggestion
            'Do we have your permission to collect usage data and metrics to help us improve CodeQL for VSCode? See [TELEMETRY.md](https://github.com/github/vscode-codeql/blob/main/.github/TELEMETRY.md) for details of what we collect and how we use it.',
```

File: TELEMETRY.md
Comment (adityasharad) [inline TELEMETRY.md:9]: As discussed separately, there are a few more anonymous fields that @aeisenberg will either delete or mention here.

**Review by adityasharad — COMMENTED** [review]
Code looks fine, minor suggestions on text for consistency, and a question about the release workflow changes.

File: .github/TELEMETRY.md
Comment (aeisenberg) [reply .github/TELEMETRY.md]: Some of this is deleted. I'll add the rest.

File: .github/TELEMETRY.md
Comment (aeisenberg) [reply .github/TELEMETRY.md]: As mentioned during our chat, not going to change this. Location is not collected. Rather, application insights uses a service to guess the location based on the IP address.

File: .github/workflows/release.yml
Comment (aeisenberg) [reply .github/workflows/release.yml:100]: Darn...must be something about the merging I've been doing to keep it up to date with `main`. This should already be in main, but I'll check to make sure.

Comment (adityasharad) [reply .github/workflows/release.yml:100]: I think it may be worth rebasing this branch for a clean diff, rather than merging in main. Your changes to the release workflow on main are more recent.

Comment (aeisenberg) [reply .github/workflows/release.yml:100]: Actually, these changes were performed as part of the telemetry change in order to handle the APPLICATION_INSIGHTS key. It's not directly related to telemetry, but I cleaned up the yml as part of it.

It's a bit of a messy rebase and one that I've half performed several times already. Probably in the end, I should just squash this all into a single commit.

File: extensions/ql-vscode/TELEMETRY.md
Comment (sj) [inline extensions/ql-vscode/TELEMETRY.md:37]: Could you add here that data is only collected if you choose to opt in? That's just to make sure that people understand that they don't have to do anything if they don't want their data collected.

**Review by sj — CHANGES_REQUESTED** [review]

**Comment by sj** [comment]
@aeisenberg : I think some changes that you made previously have now disappeared again. I've added some more comments in places where I think this might have happened (or I might have just completely misremembered). Thanks!

File: extensions/ql-vscode/TELEMETRY.md
Comment (aeisenberg) [reply extensions/ql-vscode/TELEMETRY.md:3]: Well, I can't say "anonymous" because strictly speaking IP address and installation id is not anonymous. Even though we create the id, it is theoretically possible to map back to a single user. That's why I am separating "anonymous" data from IP address and installation id.

File: extensions/ql-vscode/README.md
Comment (aeisenberg) [reply extensions/ql-vscode/README.md:116]: Same as above...

File: extensions/ql-vscode/TELEMETRY.md
Comment (sj) [reply extensions/ql-vscode/TELEMETRY.md:3]: Yeah, I think you're right. And I can't really think of a more accurate/appropriate word, so let's leave it as is.

File: TELEMETRY.md
Comment (aeisenberg) [reply TELEMETRY.md]: Done.

File: extensions/ql-vscode/src/telemetry.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/telemetry.ts]: Done.

**Review by sj — APPROVED** [review]

**Pull request merged**
